src/train.py

```diff
@@ -0,0 +1,108 @@
+    model = lstm_probing(input_size=tokenizer_length, hidden_size=encoder_states.shape[2], device=args.device, 
+    output_size=tokenizer_length)
+    model = model.to(args.device)
+    optimizer = optim.Adam(model.parameters(), lr=args.learning_rate)
+    criterion = nn.CrossEntropyLoss()
+    print_loss_total = 0 
+    plot_loss_total = 0 
+    plot_losses = []
+    plot_perplexity = []
+    plot_epochs = []
+    start = time.time()
+
+    for epoch in range(1, args.max_epochs + 1):
+        loss, perplexity = train(model, dataloader, criterion, encoder_states, epoch, optimizer, args)
+        print_loss_total += loss
+        plot_loss_total += loss
+
+        if epoch % args.print_every == 0:
+            print_loss_avg = print_loss_total / args.print_every
+            print_loss_total = 0
+            print('%s (%d %d%%) %.4f' % (util.timeSince(start, epoch / args.max_epochs),
+                                         epoch, epoch / args.max_epochs * 100, print_loss_avg))
+
+        if epoch % args.plot_every == 0:
+            plot_loss_avg = plot_loss_total / args.plot_every
+            plot_losses.append(plot_loss_avg)
+            plot_perplexity.append(np.exp(plot_loss_avg))
+            plot_loss_total = 0
+            plot_epochs.append(epoch)
+    
+    return model, dataloader
+
+
+def train(model, dataloader, criterion, encoder_states, epoch, optimizer, args):
+    model.train()
+    # x dimension: (batch, 1, seq_lenth=768)
+    # y dimension: (batch, seq_lenth=55)
+    for batch, (x, y) in enumerate(dataloader):
+        x_tensor = torch.stack(list(x), dim=0).to(args.device)  # (batch, 768)
+        y_tensor = torch.stack(y, dim=0).to(args.device)  # (batch, 55)
+
+        optimizer.zero_grad()
+
+        hidden = model.init_hidden(x_tensor)  # (batch, 768), (batch, 768)
+        hidden = (hidden[0].unsqueeze(0), hidden[1].unsqueeze(0))
+        y_lengths = (y_tensor != 50256).long().sum(-1)
+        packed_logits, packed_targets = model.forward(y_tensor, y_lengths, hidden)
+
+        loss = criterion(packed_logits, packed_targets)
+        
+        loss.backward()
+        optimizer.step()
+
+        print({ 'epoch': epoch, 'loss': loss.item(), 'perplexity': torch.exp(loss) })
+        return loss.item(), torch.exp(loss)
+
+@hydra.main(config_path="conf", config_name="train_probing_config")
+def main(args: DictConfig) -> None:
+    tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
+    tokenizer_model = GPT2Model.from_pretrained("gpt2")
+    explanation_dataset = data.load_explanation(args, tokenizer)
+    for split in args.data.splits:
+        print("======"+split+"======")
+        encoder_states = data.load_encoder_states(args, split)
+        exp_dataset = explanation_dataset[split]
+        if split == "train":
+            model, dataloader = train_iterations(encoder_states, exp_dataset, len(tokenizer), split, args)
```

Comment:
As I mentioned previously I think it's important to monitor the validation BLEU while training your model, and train until it has peaked. It seems like right now you're training for a fixed number of epochs (60) before then validating the model. You should wrap the behavior in lines 95-105 in some function that you can call to validate your model as it's training.

(There'll be no need to save the model to disk and reload it, since you can just call validate on the model itself while it's training. Be sure to set `model.eval()` before validation (and set `model.train()` during training))

This is called [**early stopping**](https://en.wikipedia.org/wiki/Early_stopping): you monitor performance on some metric you care about (validation accuracy for held out data) and stop training when metric performance hasn't improved, keeping the last model state that resulted in the highest performance. In order to implement this, you can simply copy the model state dict when you want to keep it around in memory (e.g. via `best_model = copy.deepcopy(model.state_dict())`). You can then restore this state with `model.load_state_dict(best_model)` and save that state to disk, instead of the current model state.

Reply:
updated